azkaban-common/src/main/java/azkaban/executor/container/ContainerImplUtils.java

```diff
@@ -82,4 +93,82 @@ public static int getFlowNameHashValMapping(ExecutableFlow flow) {
     int flowNameHashVal = Math.abs(MurmurHash3.hash32x86(flowNameBytes));
     return flowNameHashVal % 100 + 1;
   }
-}
+
+  public static Set<String> getProxyUsersForFlow(final ProjectManager projectManager,
+      final ExecutableFlow flow) {
+    final Set<String> proxyUsers = new HashSet<>();
+    // Get the project and flow Object that needs to be used repeatedly in the DAG.
+    Project project = projectManager.getProject(flow.getProjectId());
+    Flow flowObj = project.getFlow(flow.getFlowId());
+
+    /* Get the flow properties and check if the proxy user is present in the highest level of the
+     flow and not the job. Passing null as the job name is able to get us the top level flow
+     properties mentioned for the flow.*/
+
+    Props flowProps = projectManager.getProperties(project, flowObj,
+        null, flow.getJobSource());
+    if (flowProps != null) {
+      String proxyUserFromFlowProp = flowProps.getString(USER_TO_PROXY, null);
+      if (proxyUserFromFlowProp != null && !proxyUserFromFlowProp.isEmpty()) {
+        proxyUsers.add(proxyUserFromFlowProp);
+      }
+    }
+    // DFS Walk of the Graph to find all the Proxy Users.
+    populateProxyUsersForFlow(flow, flow, flowObj, project, projectManager, proxyUsers);
+    return proxyUsers;
+  }
+
+  public static void populateProxyUsersForFlow(final ExecutableFlow flow,
+      final ExecutableNode node, final Flow flowObj, final Project project,
+      final ProjectManager projectManager, Set<String> proxyUsers) {
+    if (node instanceof ExecutableFlowBase) {
+      final ExecutableFlowBase base = (ExecutableFlowBase) node;
+      for (ExecutableNode subNode : base.getExecutableNodes()) {
+        populateProxyUsersForFlow(flow, subNode, flowObj, project, projectManager, proxyUsers);
+      }
+    } else {
+      // If a node is disabled, we don't need to initialize its jobType container image when
+      // creating a pod.
+      if (node.getStatus() != Status.DISABLED) {
+        Props currentNodeProps = projectManager.getProperties(project, flowObj,
+            node.getId(), node.getJobSource());
+        // Get the node level property for proxy user.
+        String userToProxyFromNode = (currentNodeProps != null) ? currentNodeProps.getString(USER_TO_PROXY, null) : null;
```

Comment:
Please consider defaulting to empty string here and in line 140 below to avoid unnecessary null checks and improve readability.